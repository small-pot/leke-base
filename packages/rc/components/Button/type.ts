declare const ButtonTypes: ['main','default','secondary','dashed','link'];
export declare type ButtonType = typeof ButtonTypes[number];
declare const ButtonSizes:['large','middle','small'];
export declare type ButtonSize = typeof ButtonSizes[number];
declare const ButtonShapes:['circle','round'];
export declare type ButtonShape = typeof ButtonShapes[number];
declare const ButtonHTMLTypes:['submit','button','reset'];
export declare type ButtonHTMLType = typeof ButtonHTMLTypes[number];