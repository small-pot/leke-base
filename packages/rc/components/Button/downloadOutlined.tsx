import React, { ReactElement } from 'react';
import {Upload} from '@leke/icons';

interface Props {
    
}

export default function downloadOutlined({}: Props): ReactElement {
    return (
        <span style={{verticalAlign:'text-top'}}>
            <Upload />
        </span>
    );
}
