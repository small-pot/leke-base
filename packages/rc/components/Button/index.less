@import "../../less/theme.less";
@import "./mixins.less";

@normal-color:#8c8c8c;
@normal-hover-color:#bfbfbf;
@normal-click-color:#595959;
@leke-btn-disabled:#bbe9df;

// 警示按钮

@keyframes spin {
    from {
      transform: rotate(0deg);
    }
    to {
      transform: rotate(360deg);
    }
}

.leke-btn {
    position: relative;
    display: inline-block;
    font-weight: 400;
    white-space: nowrap;
    text-align: center;
    background-image: none;
    border: 1px solid transparent;
    outline: 0;
    height: 32px;
    padding: 0 15px;
    font-size: 14px;
    border-radius: 4px;
    cursor: pointer;
    &-large{
        .type(104px,40px,16px)
    }
    &-middle{
        .type(84px,32px,14px)
    }
    &-small{
        .type(68px,28px,12px)
    }
    &-icon-only{
        width: 32px;
        height: 32px;
        padding: 2.4px 0;
        font-size: 16px;
        &.leke-btn-large{
            width: 40px;
            height: 40px;
            font-size: 16px;
        }
        &.leke-btn-middle{
            width: 32px;
            height: 32px;
            font-size: 14px;
        }
        &.leke-btn-small{
            width: 28px;
            height: 28px;
            font-size: 12px;
        }
        & .leke-btn-content{
            margin-left: 8px;
        }
    }
    &-circle{
        &.leke-btn-large{
            width: 40px;
            height: 40px;
            padding: 4.9px 0;
            font-size: 18px;
            min-width:40px;
            border-radius: 50%;
            text-align: center
        }
        &.leke-btn-middle{
            min-width:32px;
            border-radius: 50%;
            text-align: center
        }
        &.leke-btn-small{
            min-width:28px;
            border-radius: 50%;
            text-align: center
        }
    }
    &-loading{
        position: relative;
        &::before{
            display: block;
            position: absolute;
            top: -1px;
            right: -1px;
            bottom: -1px;
            left: -1px;
            z-index: 1;
            background: #fff;
            opacity: 0.35;
            border-radius: inherit;
            transition: opacity 0.2s;
            content: "";
            pointer-events: none;
        }
        & .leke-btn-content{
            margin-left: 8px;
        }
    }
    &-block{
        width: 100%;
    }
    &-loading-icon{
        width: 18px;
        height: 18px;
        background:url('./assets/loading.svg');
        display: inline-block;
        background-size: cover;
        animation: spin 1s linear infinite;
        vertical-align: sub;
    }
    &-content{
        // vertical-align: middle;
    }
    &-default{
        .btn-default(@base-color,@hover-color,@click-color);
    }
    &-dashed{
        .btn-dashed(@base-color,@hover-color,@click-color)
    }
    &-normal{
        .btn-normal(@normal-color,@normal-hover-color,@normal-click-color)
    }
    &-text {
        .btn-text(@base-color,@hover-color,@click-color)
    }
    &-primary {
        .btn-primary(@base-color,@hover-color,@click-color)
    }
    &-danger&-primary{
        .btn-dangerours('primary')
    }
    &-danger&-text{
        .btn-dangerours('text')
    }
    &-danger&-dashed{
        .btn-dangerours('dashed')
    }
    &-danger&-default{
        .btn-dangerours('default')
    }
    &-warning&-primary{
        .btn-warning('primary')
    }
    &-warning&-text{
        .btn-warning('text')
    }
    &-warning&-dashed{
        .btn-warning('dashed')
    }
    &-warning&-default{
        .btn-warning('default')
    }
}
.leke-content{
    display: inline-block;
}
.leke-round{
    border-radius: 40px;
}
.leke-btnBackgroundGhost {
    color: #fff;
    border-color: #fff;
}
.leke-btnBackgroundGhost.leke-btn-primary {
    background: #bbe9df;
    border: none;
    color: @base-color;
    text-shadow: none;
}
.leke-btnBackgroundGhost.leke-btn-primary:hover {
    background-color: #fff;
}

.leke-btnBackgroundGhost.leke-btn-default {
    background: transparent;
    border: 1px solid #fff;
    color: #fff;
    text-shadow: none;
}

.leke-btnBackgroundGhost.leke-btn-default:hover {
    color: @base-color;
    background-color: #fff;
}
.leke-btnBackgroundGhost.leke-btn-dashed {
    background: transparent;
    border: 1px dashed #fff;
    color: #fff;
    text-shadow: none;
}

.leke-btnBackgroundGhost.leke-btn-dashed:hover {
    color: @base-color;
    border: none;
    background-color: #fff;
}

.leke-abtnBackgroundGhost.leke-btn-primary[disabled],
.leke-abtnBackgroundGhost.leke-btn-primary[disabled]:hover {
    background-color:#fff;
    color: #cccccc;
    cursor: auto;
}

.leke-btnDisabled.leke-btn-primary[disabled]:hover,
.leke-btnDisabled.leke-btn-primary[disabled]{
    background-color: @leke-btn-disabled;
    .btnDisabled(#fff);
}
.leke-btnDisabled.leke-btn-default[disabled]:hover,
.leke-btnDisabled.leke-btn-default[disabled]{
    border: 1px solid @leke-btn-disabled;
    .btnDisabled(@leke-btn-disabled);
}
.leke-btnDisabled.leke-btn-dashed[disabled]:hover,
.leke-btnDisabled.leke-btn-dashed[disabled]{
    border: 1px dashed @leke-btn-disabled;
    .btnDisabled(@leke-btn-disabled);
}
.leke-btnDisabled.leke-btn-text[disabled]:hover,
.leke-btnDisabled.leke-btn-text[disabled]{
    .btnDisabled(@leke-btn-disabled);
}


.leke-btn-primary[disabled]:hover,
.leke-btn-primary[disabled]{
    background-color: #d9d9d9;
    .btnDisabled(#fff);
}
.leke-btn-default[disabled]:hover,
.leke-btn-default[disabled]{
    border: 1px solid #d9d9d9;
    .btnDisabled(#d9d9d9);
}
.leke-btn-normal[disabled]:hover,
.leke-btn-normal[disabled]{
    border: 1px solid #d9d9d9;
    .btnDisabled(#d9d9d9);
}
.leke-btn-dashed[disabled]:hover,
.leke-btn-dashed[disabled]{
    border: 1px dashed #d9d9d9;
    .btnDisabled(#d9d9d9);
}
.leke-btn-text[disabled]:hover,
.leke-btn-text[disabled]{
    .btnDisabled(#d9d9d9);
}
