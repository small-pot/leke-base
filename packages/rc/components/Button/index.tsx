import React, { ReactNode } from 'react';
import classNames from 'classnames';
import { ButtonShape, ButtonSize, ButtonType } from './type';

export type Props = {
    children?: ReactNode;
    type?: ButtonType;
    ghost?: boolean;
    disabled?: boolean;
    className?: string;
    size?:ButtonSize;
    shape?:ButtonShape;
    icon?:ReactNode;
    loading?:boolean;
    lekeDisabled?:boolean;
    block?:boolean;
    danger?:boolean;
    warning?:boolean;
    onClick?: () => any;
};

const Button = (props: Props) => {
    const { type, ghost, disabled, className,size,shape,icon,children,loading,block,lekeDisabled,danger,warning, onClick: fn } = props;

    const classes = classNames(className, {
        'leke-btn': true,
        'leke-btn-danger':danger,
        'leke-btn-warning':warning,
        [type ? `leke-btn-${type}` : 'leke-btn-default']: true,
        [shape ? `leke-btn-${shape}`:'' ]:true,
        [size ? `leke-btn-${size}`:'leke-btn-middle']:true,
        'leke-btn-loading':loading,
        'leke-btn-icon-only':icon,
        'leke-btn-block':block,
        'leke-btnBackgroundGhost': ghost,
        'leke-btnDisabled':lekeDisabled,
    });

    const LoadingIcon = () => {
        return <div className="leke-btn-loading-icon" />;
    };

    const iconNode = loading ? <LoadingIcon /> : icon  ? icon : null;

    const kids = children ? 
        <span className='leke-btn-content'>{children}</span>
        : 
        null;

    return (
        <button disabled={disabled || lekeDisabled} onClick={() => fn && fn()} className={classes}>
            {iconNode}
            {kids}
        </button>
    );
};

export default Button;
