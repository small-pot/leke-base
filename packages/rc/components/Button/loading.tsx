import React, { ReactElement } from 'react';


export default function loading(): ReactElement {
    return <div className="leke-btn-loading-icon" />;
}
