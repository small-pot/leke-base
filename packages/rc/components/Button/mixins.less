@btn-primary:'primary';
// 失败按钮
@danger-btn-color:#ff4848;
@danger-btn-hover-color:#ff7571;
@danger-btn-click-color:#d93337;
@leke-danger-btn-disabled:#ffc8c2;
//警示按钮
@warning-btn-color:#ff9900;
@warning-btn-hover-color:#ffb029;
@warning-btn-click-color:#d97b00;
@leke-warning-btn-disabled:#ffe7a3;

.type(@width,@height,@fontSize){
    min-width: @width;
    height: @height;
    font-size: @fontSize;
}
.btnDisabled(@color){
    color: @color;
    cursor: not-allowed;
}
.btn-primary(@color,@hoverColor,@clickColor){
    color: #fff;
    border: none;
    background-color: @color;
    &:hover, 
    &:focus{
        background-color: @hoverColor;
    }
    &:active{
        background-color: @clickColor;
    }
}
.btn-default(@color,@hoverColor,@clickColor){
    color: @color;
    background: transparent;
    border: 1px solid @color;
    &:hover, 
    &:focus{
        color: @hoverColor;
        border-color: @hoverColor;
    }
    &:active{
        color: @clickColor;
        border-color: @clickColor;
    }
}
.btn-normal(@color,@hoverColor,@clickColor){
    color: @color;
    background: transparent;
    border: solid 1px @color;
    &:hover, 
    &:focus{
        color: @hoverColor;
        border-color: @hoverColor;
    }
    &:active{
        color: @clickColor;
        border-color: @clickColor;
    }
}
.btn-dashed(@color,@hoverColor,@clickColor){
    color: @color;
    background: transparent;
    border: 1px dashed @color;
    &:hover, 
    &:focus{
        color: @hoverColor;
        border-color: @hoverColor;
    }
    &:active{
        color: @clickColor;
        border-color: @clickColor;
    }
}
.btn-text(@color,@hoverColor,@clickColor){
    color: @color;
    background-color: transparent;
    border-color: transparent;
    box-shadow: none;
    &:hover, 
    &:focus{
        color: @hoverColor;
    }
    &:active{
        color: @clickColor;
    }
}
.btn-dangerours(@type){
    & when(@type = 'primary'){
        .btn-primary(@danger-btn-color,@danger-btn-hover-color,@danger-btn-click-color);
        &.leke-btnDisabled.leke-btn-primary[disabled]:hover,
        &.leke-btnDisabled.leke-btn-primary[disabled]{
            background-color: @leke-danger-btn-disabled;
            .btnDisabled(#fff);
        }
    }
    & when(@type = 'text'){
        .btn-text(@danger-btn-color,@danger-btn-hover-color,@danger-btn-click-color);
        &.leke-btnDisabled.leke-btn-text[disabled]:hover,
        &.leke-btnDisabled.leke-btn-text[disabled]{
            .btnDisabled(@leke-danger-btn-disabled);
        }
    }
    & when(@type = 'dashed'){
        .btn-dashed(@danger-btn-color,@danger-btn-hover-color,@danger-btn-click-color);
        &.leke-btnDisabled.leke-btn-dashed[disabled]:hover,
        &.leke-btnDisabled.leke-btn-dashed[disabled]{
            border: 1px dashed @leke-danger-btn-disabled;
            .btnDisabled(@leke-danger-btn-disabled);
        }
    }
    & when(@type = 'default'){
        .btn-default(@danger-btn-color,@danger-btn-hover-color,@danger-btn-click-color);
        &.leke-btnDisabled.leke-btn-default[disabled]:hover,
        &.leke-btnDisabled.leke-btn-default[disabled]{
            border: 1px solid @leke-danger-btn-disabled;
            .btnDisabled(@leke-danger-btn-disabled);
        }
    }
}
.btn-warning(@type){
    & when(@type = 'primary'){
        .btn-primary(@warning-btn-color,@warning-btn-hover-color,@warning-btn-click-color);
        &.leke-btnDisabled.leke-btn-primary[disabled]:hover,
        &.leke-btnDisabled.leke-btn-primary[disabled]{
            background-color: @leke-warning-btn-disabled;
            .btnDisabled(#fff);
        }
    }
    & when(@type = 'text'){
        .btn-text(@warning-btn-color,@warning-btn-hover-color,@warning-btn-click-color);
        &.leke-btnDisabled.leke-btn-text[disabled]:hover,
        &.leke-btnDisabled.leke-btn-text[disabled]{
            .btnDisabled(@leke-warning-btn-disabled);
        }
    }
    & when(@type = 'dashed'){
        .btn-dashed(@warning-btn-color,@warning-btn-hover-color,@warning-btn-click-color);
        &.leke-btnDisabled.leke-btn-dashed[disabled]:hover,
        &.leke-btnDisabled.leke-btn-dashed[disabled]{
            border: 1px dashed @leke-warning-btn-disabled;
            .btnDisabled(@leke-warning-btn-disabled);
        }
    }
    & when(@type = 'default'){
        .btn-default(@warning-btn-color,@warning-btn-hover-color,@warning-btn-click-color);
        &.leke-btnDisabled.leke-btn-default[disabled]:hover,
        &.leke-btnDisabled.leke-btn-default[disabled]{
            border: 1px solid @leke-warning-btn-disabled;
            .btnDisabled(@leke-warning-btn-disabled);
        }
    }
}


