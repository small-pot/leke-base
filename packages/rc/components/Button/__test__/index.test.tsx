import React from "react";
import Button from "../index";
import {render,waitFor,screen, fireEvent} from '@testing-library/react';
import userEvent from "@testing-library/user-event";

describe('Button ', function() {
    it('button render type', () => {
        render(<Button type='main' />);
        expect(document.body).toMatchSnapshot();
    });
    it('button render size', () => {
        render(<Button size='large'/>);
        expect(document.body).toMatchSnapshot();
    });
    it('button render shape', () => {
        render(<Button shape='circle'/>);
        expect(document.body).toMatchSnapshot();
    });
    it('button render disabled', () => {
        render(<Button disabled/>);
        expect(document.body).toMatchSnapshot();
    });
    it('button render loading', () => {
        const {container} = render(
            <span> 1111 </span>
        );
        render(<Button icon={container} loading/>);
        expect(document.body).toMatchSnapshot();
    });
    it('button should be called', () => {
        const onClick = jest.fn();
        render( <Button onClick={onClick} >点击</Button>);
        fireEvent.click(screen.getByText('点击'));
        expect(onClick).toHaveBeenCalledTimes(1);
    });
});