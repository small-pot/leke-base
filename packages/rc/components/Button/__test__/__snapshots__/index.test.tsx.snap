// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button  button render disabled 1`] = `
<body>
  <div>
    <button
      class="leke-btn leke-btn-default leke-btn-middle"
      disabled=""
    />
  </div>
</body>
`;

exports[`Button  button render loading 1`] = `
<body>
  <div>
    <span>
       1111 
    </span>
  </div>
  <div>
    <button
      class="leke-btn leke-btn-default leke-btn-middle leke-btn-loading leke-btn-icon-only"
    >
      <div
        class="leke-btn-loading-icon"
      />
    </button>
  </div>
</body>
`;

exports[`Button  button render shape 1`] = `
<body>
  <div>
    <button
      class="leke-btn leke-btn-default leke-btn-circle leke-btn-middle"
    />
  </div>
</body>
`;

exports[`Button  button render size 1`] = `
<body>
  <div>
    <button
      class="leke-btn leke-btn-default leke-btn-large"
    />
  </div>
</body>
`;

exports[`Button  button render type 1`] = `
<body>
  <div>
    <button
      class="leke-btn leke-btn-main leke-btn-middle"
    />
  </div>
</body>
`;
